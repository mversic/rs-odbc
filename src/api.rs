use crate::handle::AsMutSQLHANDLE;
use crate::{
    env::EnvAttr, AnsiType,
    handle::{Allocate, AsSQLHANDLE, HandleIdentifier, HDBC, HENV, SQLHDBC},
    AsMutPtr, AsMutRawSlice, AsRawParts, AsRawSQLCHARSlice, AsMutRawSQLCHARSlice,
    DriverCompletion, GetAttr, OdbcAttr, SetAttr, SQLCHAR, SQLHANDLE, SQLHENV, SQLINTEGER,
    SQLPOINTER, SQLRETURN, SQLSMALLINT, SQLUSMALLINT,
};
use std::mem::MaybeUninit;

// TODO: Fix linking
// static linking is not currently supported here for windows
#[cfg_attr(windows, link(name = "odbc32"))]
#[cfg_attr(all(not(windows), not(r#static)), link(name = "odbc"))]
#[cfg_attr(all(not(windows), r#static), link(name = "odbc", kind = "static"))]
extern "system" {
    #[link_name = "SQLAllocHandle"]
    fn AllocHandle(
        HandleType: SQLSMALLINT,
        InputHandle: SQLHANDLE,
        OutputHandlePtr: *mut SQLHANDLE,
    ) -> SQLRETURN;

    #[link_name = "SQLFreeHandle"]
    pub(crate) fn FreeHandle(HandleType: SQLSMALLINT, Handle: SQLHANDLE) -> SQLRETURN;

    #[link_name = "SQLSetEnvAttr"]
    fn SetEnvAttr(
        EnvironmentHandle: HENV,
        Attribute: SQLINTEGER,
        ValuePtr: SQLPOINTER,
        StringLength: SQLINTEGER,
    ) -> SQLRETURN;

    #[link_name = "SQLGetEnvAttr"]
    fn GetEnvAttr(
        EnvironmentHandle: HENV,
        Attribute: SQLINTEGER,
        ValuePtr: SQLPOINTER,
        BufferLength: SQLINTEGER,
        StringLengthPtr: *mut SQLINTEGER,
    ) -> SQLRETURN;

    #[link_name = "SQLDriverConnectA"]
    fn DriverConnectA(
        ConnectionHandle: HDBC,
        WindowHandle: SQLHANDLE,
        InConnectionString: *const SQLCHAR,
        StringLength1: SQLSMALLINT,
        OutConnectionString: *mut SQLCHAR,
        BufferLength: SQLSMALLINT,
        StringLength2Ptr: *mut SQLSMALLINT,
        DriverCompletion: SQLUSMALLINT,
    ) -> SQLRETURN;

    #[link_name = "SQLDisconnect"]
    pub(crate) fn Disconnect(ConnectionHandle: HDBC) -> SQLRETURN;
}

#[inline]
pub fn SQLAllocHandle<'src, OH: Allocate<'src>>(
    HandleType: OH::Identifier,
    InputHandle: &'src OH::SrcHandle,
    OutputHandlePtr: &mut MaybeUninit<OH>,
) -> SQLRETURN
where
    OH::SrcHandle: AsSQLHANDLE,
    OH: AsMutSQLHANDLE,
{
    unsafe {
        AllocHandle(
            OH::Identifier::IDENTIFIER,
            // TODO: Is this UB? Handle will be modified on the extern C side but it's passed
            // here from shared reference. It has to be shared reference to be able to allocate
            // multiple children. To resolve the issue RefCell with PhantomData<Ref<'src, T>>
            // can be used, but it's hard to implement
            //
            // Also, what about multithreaded code?
            InputHandle.as_SQLHANDLE(),
            OutputHandlePtr.as_mut_ptr().cast(),
        )
    }
}

#[inline]
pub fn SQLSetEnvAttr<A: EnvAttr, T>(
    EnvironmentHandle: &mut SQLHENV,
    Attribute: A,
    ValuePtr: &T,
) -> SQLRETURN
where
    A: SetAttr<AnsiType, T>,
    T: AsRawParts<OdbcAttr, SQLINTEGER>,
{
    let ValuePtr = ValuePtr.as_raw_parts();

    unsafe {
        SetEnvAttr(
            EnvironmentHandle.as_mut_SQLHANDLE(),
            A::IDENTIFIER,
            ValuePtr.0,
            ValuePtr.1,
        )
    }
}

pub fn SQLGetEnvAttr<A: EnvAttr, T>(
    EnvironmentHandle: &SQLHENV,
    Attribute: A,
    ValuePtr: &mut T,
    StringLengthPtr: &mut MaybeUninit<T::StrLen>,
) -> SQLRETURN
where
    A: GetAttr<AnsiType, T>,
    T: AsMutRawSlice<OdbcAttr, SQLINTEGER>,
    MaybeUninit<T::StrLen>: AsMutPtr<SQLINTEGER>,
{
    let ValuePtr = ValuePtr.as_mut_raw_slice();

    unsafe {
        GetEnvAttr(
            EnvironmentHandle.as_SQLHANDLE(),
            A::IDENTIFIER,
            ValuePtr.0,
            ValuePtr.1,
            StringLengthPtr.as_mut_ptr().cast(),
        )
    }
}

pub fn SQLDriverConnectA<
    C: AsRawSQLCHARSlice<SQLSMALLINT> + ?Sized,
    MC: AsMutRawSQLCHARSlice<SQLSMALLINT>,
>(
    ConnectionHandle: &mut SQLHDBC,
    WindowHandle: Option<SQLHANDLE>,
    InConnectionString: &C,
    OutConnectionString: &mut MC,
    StringLength2Ptr: &mut MaybeUninit<SQLSMALLINT>,
    DriverCompletion: DriverCompletion,
) -> SQLRETURN {
    let InConnectionString = InConnectionString.as_raw_SQLCHAR_slice();
    let OutConnectionString = OutConnectionString.as_mut_raw_SQLCHAR_slice();

    unsafe {
        let res = DriverConnectA(
            ConnectionHandle.as_mut_SQLHANDLE(),
            std::ptr::null_mut(),
            InConnectionString.0,
            InConnectionString.1,
            OutConnectionString.0,
            OutConnectionString.1,
            StringLength2Ptr.as_mut_ptr(),
            DriverCompletion as SQLUSMALLINT,
        );

        res
    }
}

pub fn SQLDisconnect(ConnectionHandle: &mut SQLHDBC) -> SQLRETURN {
    unsafe { Disconnect(ConnectionHandle.as_mut_SQLHANDLE()) }
}
