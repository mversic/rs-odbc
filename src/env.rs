use crate::attr::{Attr, AttrGet, AttrLen, AttrSet};
use crate::str::{OdbcChar, OdbcStr};
use crate::{Ident, OdbcDefined, SQLCHAR, SQLINTEGER, SQLUINTEGER, SQLWCHAR};
use rs_odbc_derive::{odbc_type, Ident};
use std::mem::MaybeUninit;

pub trait EnvAttr<A: Ident, V: OdbcVersion>:
    Attr<A, DefinedBy = OdbcDefined> + AttrLen<OdbcDefined, SQLINTEGER>
{
}

// Implement EnvAttr for all versions of environment attributes
impl<A: Ident, T: Ident> EnvAttr<A, SQL_OV_ODBC3_80> for T where T: EnvAttr<A, SQL_OV_ODBC3> {}
impl<A: Ident, T: Ident> EnvAttr<A, SQL_OV_ODBC4> for T where T: EnvAttr<A, SQL_OV_ODBC3_80> {}
impl<A: Ident, CH: OdbcChar> EnvAttr<A, SQL_OV_ODBC3_80> for OdbcStr<CH> where
    OdbcStr<CH>: EnvAttr<A, SQL_OV_ODBC3>
{
}
impl<A: Ident, CH: OdbcChar> EnvAttr<A, SQL_OV_ODBC4> for OdbcStr<CH> where
    OdbcStr<CH>: EnvAttr<A, SQL_OV_ODBC3_80>
{
}

// Implement EnvAttr for uninitialized environment attributes
impl<A: Ident, T: Ident, V: OdbcVersion> EnvAttr<A, V> for MaybeUninit<T>
where
    T: EnvAttr<A, V>,
    Self: AttrLen<OdbcDefined, SQLINTEGER>,
{
}
impl<A: Ident, V: OdbcVersion> EnvAttr<A, V> for OdbcStr<MaybeUninit<SQLCHAR>> where
    OdbcStr<SQLCHAR>: EnvAttr<A, V>
{
}
impl<A: Ident, V: OdbcVersion> EnvAttr<A, V> for OdbcStr<MaybeUninit<SQLWCHAR>> where
    OdbcStr<SQLWCHAR>: EnvAttr<A, V>
{
}

// Implement EnvAttr for references to character environment attributes (used by AttrSet)
impl<A: Ident, CH: OdbcChar, V: OdbcVersion> EnvAttr<A, V> for &OdbcStr<CH>
where
    OdbcStr<CH>: EnvAttr<A, V>,
    Self: AttrSet<A>,
{
}

//=====================================================================================//
//-------------------------------------Attributes--------------------------------------//

// TODO: Consider using const generics for OdbcVersion once it's available on stable,
// otherwise don't expose this attribute unless there is a valid use-case
#[derive(Ident)]
#[identifier(SQLINTEGER, 200)]
#[allow(non_camel_case_types)]
// This is read-only attribute because
// it's handled by the type system
pub(crate) struct SQL_ATTR_ODBC_VERSION;
//unsafe impl Attr<SQL_ATTR_ODBC_VERSION> for OdbcVersion {
//    type DefinedBy = OdbcDefined;
//}
//impl EnvAttr<SQL_ATTR_ODBC_VERSION, SQL_OV_ODBC3> for OdbcVersion {}
//unsafe impl AttrGet<SQL_ATTR_ODBC_VERSION> for OdbcVersion {}

#[derive(Ident)]
#[identifier(SQLINTEGER, 202)]
#[allow(non_camel_case_types)]
pub struct SQL_ATTR_CP_MATCH;
unsafe impl Attr<SQL_ATTR_CP_MATCH> for CpMatch {
    type DefinedBy = OdbcDefined;
}
impl EnvAttr<SQL_ATTR_CP_MATCH, SQL_OV_ODBC3> for CpMatch {}
unsafe impl AttrGet<SQL_ATTR_CP_MATCH> for CpMatch {}
unsafe impl AttrSet<SQL_ATTR_CP_MATCH> for CpMatch {}

#[derive(Ident)]
#[identifier(SQLINTEGER, 201)]
#[allow(non_camel_case_types)]
pub struct SQL_ATTR_CONNECTION_POOLING;
unsafe impl Attr<SQL_ATTR_CONNECTION_POOLING> for ConnectionPooling {
    type DefinedBy = OdbcDefined;
}
impl EnvAttr<SQL_ATTR_CONNECTION_POOLING, SQL_OV_ODBC3_80> for ConnectionPooling {}
unsafe impl AttrGet<SQL_ATTR_CONNECTION_POOLING> for ConnectionPooling {}
unsafe impl AttrSet<SQL_ATTR_CONNECTION_POOLING> for ConnectionPooling {}

//=====================================================================================//

pub trait OdbcVersion {
    const IDENTIFIER: SQLUINTEGER;
}
#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum SQL_OV_ODBC3 {}
impl OdbcVersion for SQL_OV_ODBC3 {
    const IDENTIFIER: SQLUINTEGER = 3;
}
#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum SQL_OV_ODBC3_80 {}
impl OdbcVersion for SQL_OV_ODBC3_80 {
    const IDENTIFIER: SQLUINTEGER = 380;
}
#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum SQL_OV_ODBC4 {}
impl OdbcVersion for SQL_OV_ODBC4 {
    const IDENTIFIER: SQLUINTEGER = 400;
}

#[odbc_type(SQLUINTEGER)]
pub struct CpMatch;
pub const SQL_CP_STRICT_MATCH: CpMatch = CpMatch(0);
pub const SQL_CP_RELAXED_MATCH: CpMatch = CpMatch(1);

#[odbc_type(SQLUINTEGER)]
pub struct ConnectionPooling;
pub const SQL_CP_OFF: ConnectionPooling = ConnectionPooling(0);
pub const SQL_CP_ONE_PER_DRIVER: ConnectionPooling = ConnectionPooling(1);
pub const SQL_CP_ONE_PER_HENV: ConnectionPooling = ConnectionPooling(2);
pub const SQL_CP_DRIVER_AWARE: ConnectionPooling = ConnectionPooling(3);
