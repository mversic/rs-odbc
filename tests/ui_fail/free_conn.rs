use rs_odbc::api::Allocate;
use rs_odbc::env::SQL_OV_ODBC3_80;
use rs_odbc::handle::{SQLHDBC, SQLHENV, SQL_NULL_HANDLE};

fn get_env_handle() -> SQLHENV<SQL_OV_ODBC3_80> {
    let (env, _) = SQLHENV::SQLAllocHandle(&SQL_NULL_HANDLE);
    env.unwrap()
}

fn main() {
    let env = get_env_handle();
    let (conn, _) = SQLHDBC::SQLAllocHandle(&env);
    let conn = conn.unwrap();

    env.SQLFreeHandle();
    conn.SQLFreeHandle();
}
