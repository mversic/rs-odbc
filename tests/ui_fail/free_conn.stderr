error[E0505]: cannot move out of `env` because it is borrowed
  --> tests/ui_fail/free_conn.rs:15:5
   |
12 |     let (conn, _) = SQLHDBC::SQLAllocHandle(&env);
   |                                             ---- borrow of `env` occurs here
...
15 |     env.SQLFreeHandle();
   |     ^^^ move out of `env` occurs here
16 |     conn.SQLFreeHandle();
   |     ---- borrow later used here
